File: examples/showcase/jira_fronting/custom-code-bundles/jira_fronting_java/src/main/java/{{ANIP_JAVA_PACKAGE_PATH}}/BackendAdapter.java
package {{ANIP_JAVA_PACKAGE_NAME}};

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import dev.anip.service.InvocationContext;

import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Base64;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.StringJoiner;

@FunctionalInterface
public interface BackendAdapter {

    Map<String, Object> execute(Map<String, Object> capability, Map<String, Object> plan, Map<String, Object> adapterInput, InvocationContext context);

    static BackendAdapter defaultAdapter() {
        return new JiraBackendAdapter();
    }
}

final class JiraBackendAdapter implements BackendAdapter {
    private static final ObjectMapper MAPPER = new ObjectMapper();
    private static final HttpClient HTTP = HttpClient.newHttpClient();

    @Override
    public Map<String, Object> execute(Map<String, Object> capability, Map<String, Object> plan, Map<String, Object> params, InvocationContext context) {
        List<String> unresolved = stringList(plan.get("unresolved_required_backend_inputs"));
        if (!unresolved.isEmpty()) {
            return result(capability, plan, "backend_input_incomplete", Map.of("unresolved_required_backend_inputs", unresolved));
        }
        String id = text(capability.get("capability_id"));
        JiraConfig config = JiraConfig.read();
        return switch (id) {
            case "jira.backlog.search_context" -> searchBacklog(capability, plan, params, config);
            case "jira.issue.get_context" -> getIssueContext(capability, plan, params, config);
            case "jira.release_notes.prepare" -> prepareReleaseNotes(capability, plan, params, config);
            case "jira.incident_bug.prepare" -> prepareIssueCreate(capability, plan, params, "Bug");
            case "jira.story.prepare" -> prepareIssueCreate(capability, plan, params, "Story");
            case "jira.subtask.prepare" -> prepareSubtask(capability, plan, params, config);
            case "jira.customer_escalation.comment.prepare" -> preview(capability, plan, "add_comment", Map.of("method", "POST", "path", "/rest/api/3/issue/" + text(params.get("issue_key")) + "/comment", "body", Map.of("body", adf("[" + text(params.get("comment_purpose")) + "] " + text(params.get("context"))), "visibility", firstNonEmpty(text(params.get("visibility")), "internal"))), Map.of("issue_key", params.get("issue_key"), "visibility", firstNonEmpty(text(params.get("visibility")), "internal"), "comment_purpose", params.get("comment_purpose")));
            case "jira.workflow_transition.request" -> preview(capability, plan, "transition_issue", Map.of("method", "POST", "path", "/rest/api/3/issue/" + text(params.get("issue_key")) + "/transitions", "body", Map.of("transition", Map.of("id", params.get("target_status")))), Map.of("issue_key", params.get("issue_key"), "target_status", params.get("target_status")));
            case "jira.sprint_move.request" -> preview(capability, plan, "move_issues_to_sprint", Map.of("method", "POST", "path", "/rest/agile/1.0/sprint/" + text(params.get("target_sprint")) + "/issue", "body", Map.of("issues", listValue(params.get("issue_keys")))), Map.of("issue_keys", listValue(params.get("issue_keys")), "target_sprint", params.get("target_sprint")));
            case "jira.assignee_change.request" -> preview(capability, plan, "assign_issue", Map.of("method", "PUT", "path", "/rest/api/3/issue/" + text(params.get("issue_key")) + "/assignee", "body", Map.of("accountId", params.get("assignee_ref"))), Map.of("issue_key", params.get("issue_key"), "assignee_ref", params.get("assignee_ref")));
            case "jira.issue_link.request" -> preview(capability, plan, "link_issues", Map.of("method", "POST", "path", "/rest/api/3/issueLink", "body", Map.of("type", Map.of("name", params.get("link_type")), "inwardIssue", Map.of("key", params.get("source_issue_key")), "outwardIssue", Map.of("key", params.get("target_issue_key")), "comment", Map.of("body", adf(params.get("reason"))))), Map.of("requested_link_type", params.get("link_type")));
            default -> result(capability, plan, "backend_execution_stub", Map.of("note", "No Jira custom handler is registered for this capability."));
        };
    }

    private static Map<String, Object> searchBacklog(Map<String, Object> capability, Map<String, Object> plan, Map<String, Object> params, JiraConfig config) {
        String jql = issueQuery(text(params.get("project_key")), text(params.get("query")));
        if (!text(params.get("issue_type")).isEmpty()) jql += " AND issuetype = \"" + safeJql(params.get("issue_type")) + "\"";
        if (!text(params.get("status")).isEmpty()) jql += " AND status = \"" + safeJql(params.get("status")) + "\"";
        if (config == null) return result(capability, plan, "backend_not_configured", Map.of("jql_preview", jql));
        Map<String, Object> payload = searchIssues(config, jql, boundedLimit(params.get("limit"), 25, 50), "summary,status,issuetype,project,assignee,priority");
        if (payload.containsKey("error")) return backendError(capability, plan, payload);
        List<Map<String, Object>> issues = issueSummaries(payload.get("issues"));
        return result(capability, plan, "completed", Map.of("jql", jql, "result", Map.of("issues", issues, "count", issues.size(), "is_last", payload.get("isLast"))));
    }

    private static Map<String, Object> getIssueContext(Map<String, Object> capability, Map<String, Object> plan, Map<String, Object> params, JiraConfig config) {
        String issueKey = text(params.get("issue_key"));
        if (config == null) return result(capability, plan, "backend_not_configured", Map.of("path_preview", "/rest/api/3/issue/" + issueKey));
        Map<String, Object> payload = jira(config, "GET", "/rest/api/3/issue/" + url(issueKey), Map.of("fields", "summary,status,issuetype,project,assignee,priority,description"), null);
        if (payload.containsKey("error")) return backendError(capability, plan, payload);
        return result(capability, plan, "completed", Map.of("result", issueSummary(payload)));
    }

    private static Map<String, Object> prepareReleaseNotes(Map<String, Object> capability, Map<String, Object> plan, Map<String, Object> params, JiraConfig config) {
        String releaseRef = text(params.get("release_ref"));
        String audience = firstNonEmpty(text(params.get("audience")), "internal");
        String jql = "project = \"" + safeJql(params.get("project_key")) + "\" AND " + (releaseRef.equalsIgnoreCase("unversioned") ? "fixVersion is EMPTY" : "fixVersion = \"" + safeJql(releaseRef) + "\"");
        if (!text(params.get("issue_query")).isEmpty()) jql += " AND text ~ \"" + safeJql(params.get("issue_query")) + "\"";
        jql += " ORDER BY priority DESC, updated DESC";
        List<Map<String, Object>> issues = new ArrayList<>();
        if (config != null) {
            Map<String, Object> payload = searchIssues(config, jql, boundedLimit(params.get("limit"), 20, 50), "summary,status,issuetype,project");
            if (payload.containsKey("error")) return backendError(capability, plan, payload);
            issues = issueSummaries(payload.get("issues"));
        }
        return result(capability, plan, "prepared", Map.of("jql", jql, "result", Map.of("audience", audience, "issue_count", issues.size(), "issues", issues, "draft", releaseDraft(audience, releaseRef, issues)), "note", "Prepared release notes only. No Jira mutation or publication was performed."));
    }

    private static Map<String, Object> prepareIssueCreate(Map<String, Object> capability, Map<String, Object> plan, Map<String, Object> params, String issueType) {
        Map<String, Object> fields = new LinkedHashMap<>();
        fields.put("project", Map.of("key", params.get("project_key")));
        fields.put("issuetype", Map.of("name", issueType));
        fields.put("summary", text(params.get("summary")));
        if (text(capability.get("capability_id")).equals("jira.incident_bug.prepare")) {
            fields.put("description", adf(params.get("description")));
            fields.put("priority", Map.of("name", priorityForSeverity(params.get("severity"))));
        } else {
            fields.put("description", adf("Acceptance criteria:\n" + text(params.get("acceptance_criteria"))));
            if (!text(params.get("priority")).isEmpty()) fields.put("priority", Map.of("name", capitalize(text(params.get("priority")))));
        }
        List<String> labels = labels(params.get("labels"));
        if (!labels.isEmpty()) fields.put("labels", labels);
        return preview(capability, plan, "create_issue", Map.of("method", "POST", "path", "/rest/api/3/issue", "body", Map.of("fields", fields)), Map.of("project_key", params.get("project_key"), "requested_issue_type", issueType));
    }

    private static Map<String, Object> prepareSubtask(Map<String, Object> capability, Map<String, Object> plan, Map<String, Object> params, JiraConfig config) {
        Map<String, Object> fields = new LinkedHashMap<>();
        fields.put("parent", Map.of("key", params.get("parent_issue_key")));
        fields.put("issuetype", Map.of("name", "Sub-task"));
        fields.put("summary", text(params.get("summary")));
        fields.put("description", adf(params.get("description")));
        if (config != null) {
            Map<String, Object> parent = jira(config, "GET", "/rest/api/3/issue/" + url(text(params.get("parent_issue_key"))), Map.of("fields", "project"), null);
            String projectKey = nested(parent, "fields", "project", "key");
            if (!projectKey.isEmpty()) fields.put("project", Map.of("key", projectKey));
            if (!text(parent.get("id")).isEmpty()) fields.put("parent", Map.of("id", parent.get("id")));
        }
        return preview(capability, plan, "create_subtask", Map.of("method", "POST", "path", "/rest/api/3/issue", "body", Map.of("fields", fields)), Map.of("parent_issue_key", params.get("parent_issue_key")));
    }

    private static Map<String, Object> preview(Map<String, Object> capability, Map<String, Object> plan, String action, Map<String, Object> request, Map<String, Object> metadata) {
        Map<String, Object> extra = new LinkedHashMap<>();
        extra.put("approval_required", text(capability.get("operation_type")).equals("approval_gated") || text(capability.get("execution_posture")).equals("prepare_only"));
        extra.put("mutation_performed", false);
        extra.put("jira_action", action);
        extra.put("jira_request_preview", request);
        extra.put("jira_metadata", metadata);
        extra.put("note", "Prepared a governed Jira request preview. No Jira mutation was performed.");
        return result(capability, plan, "prepared", extra);
    }

    private static Map<String, Object> result(Map<String, Object> capability, Map<String, Object> plan, String status, Map<String, Object> extra) {
        Map<String, Object> result = new LinkedHashMap<>();
        result.put("execution_status", status);
        result.put("capability_id", capability.get("capability_id"));
        result.put("selected_backend", plan.get("selected_binding"));
        result.put("semantic_input", plan.get("semantic_input"));
        result.put("backend_input_contract", plan.get("backend_input_contract"));
        result.putAll(extra);
        return result;
    }

    private static Map<String, Object> backendError(Map<String, Object> capability, Map<String, Object> plan, Map<String, Object> payload) {
        return result(capability, plan, "backend_error", Map.of("jira_error", payload));
    }

    private static Map<String, Object> searchIssues(JiraConfig config, String jql, int limit, String fields) {
        return jira(config, "GET", "/rest/api/3/search/jql", Map.of("jql", jql, "maxResults", Integer.toString(limit), "fields", fields), null);
    }

    static Map<String, Object> jira(JiraConfig config, String method, String path, Map<String, String> query, Map<String, Object> body) {
        try {
            StringJoiner joiner = new StringJoiner("&");
            for (Map.Entry<String, String> entry : query.entrySet()) joiner.add(url(entry.getKey()) + "=" + url(entry.getValue()));
            URI uri = URI.create(config.baseUrl + path + (query.isEmpty() ? "" : "?" + joiner));
            HttpRequest.Builder builder = HttpRequest.newBuilder(uri).header("Accept", "application/json").header("Authorization", "Basic " + Base64.getEncoder().encodeToString((config.email + ":" + config.token).getBytes(StandardCharsets.UTF_8)));
            if (body == null) builder.method(method, HttpRequest.BodyPublishers.noBody());
            else builder.method(method, HttpRequest.BodyPublishers.ofString(MAPPER.writeValueAsString(body))).header("Content-Type", "application/json");
            HttpResponse<String> response = HTTP.send(builder.build(), HttpResponse.BodyHandlers.ofString());
            Map<String, Object> payload = response.body().isBlank() ? new LinkedHashMap<>() : MAPPER.readValue(response.body(), new TypeReference<>() {});
            if (response.statusCode() < 200 || response.statusCode() >= 300) return Map.of("error", "jira_http_error", "status", response.statusCode(), "detail", payload);
            return payload;
        } catch (Exception exc) {
            return Map.of("error", "jira_http_error", "detail", exc.toString());
        }
    }

    private static String issueQuery(String projectKey, String query) {
        String jql = "project = \"" + safeJql(projectKey) + "\"";
        if (!query.isBlank()) jql += " AND text ~ \"" + safeJql(query) + "\"";
        return jql + " ORDER BY updated DESC";
    }

    @SuppressWarnings("unchecked")
    private static List<Map<String, Object>> issueSummaries(Object value) {
        List<Map<String, Object>> result = new ArrayList<>();
        if (value instanceof List<?> items) for (Object item : items) if (item instanceof Map<?, ?> issue) result.add(issueSummary((Map<String, Object>) issue));
        return result;
    }

    private static Map<String, Object> issueSummary(Map<String, Object> issue) {
        return Map.of("key", issue.get("key"), "summary", nested(issue, "fields", "summary"), "status", nested(issue, "fields", "status", "name"), "issue_type", nested(issue, "fields", "issuetype", "name"), "project_key", nested(issue, "fields", "project", "key"), "assignee", nested(issue, "fields", "assignee", "displayName"), "priority", nested(issue, "fields", "priority", "name"));
    }

    private static Map<String, Object> adf(Object value) {
        return Map.of("type", "doc", "version", 1, "content", List.of(Map.of("type", "paragraph", "content", List.of(Map.of("type", "text", "text", text(value))))));
    }

    private static String releaseDraft(String audience, String releaseRef, List<Map<String, Object>> issues) {
        String heading = "Release " + releaseRef + " notes for " + audience;
        if (issues.isEmpty()) return heading + "\n\nNo matching Jira issues were returned for the bounded query.";
        List<String> lines = new ArrayList<>(List.of(heading, ""));
        for (Map<String, Object> issue : issues) lines.add("- " + issue.get("key") + ": " + issue.get("summary") + " (" + issue.get("status") + ")");
        return String.join("\n", lines);
    }

    private static int boundedLimit(Object value, int defaultValue, int max) {
        try {
            return Math.max(1, Math.min(Integer.parseInt(text(value)), max));
        } catch (Exception ignored) {
            return defaultValue;
        }
    }

    private static String safeJql(Object value) {
        return text(value).replace("\\", "\\\\").replace("\"", "\\\"");
    }

    private static List<String> listValue(Object value) {
        List<String> result = new ArrayList<>();
        if (value instanceof List<?> items) {
            for (Object item : items) addUnique(result, text(item));
        } else if (!text(value).isEmpty()) {
            for (String item : text(value).split(",")) addUnique(result, item.trim());
        }
        return result;
    }

    private static List<String> labels(Object value) {
        List<String> result = new ArrayList<>();
        for (String item : listValue(value)) {
            String label = item.toLowerCase().replaceAll("[^a-z0-9_.-]+", "-").replaceAll("^-+|-+$", "");
            if (!label.isEmpty()) result.add(label);
        }
        return result;
    }

    private static void addUnique(List<String> values, String value) {
        if (!value.isBlank() && !values.contains(value)) values.add(value);
    }

    private static String priorityForSeverity(Object value) {
        return switch (text(value).toLowerCase()) {
            case "sev1", "sev2" -> "High";
            case "sev4" -> "Low";
            default -> "Medium";
        };
    }

    private static String text(Object value) {
        return value == null ? "" : value.toString().trim();
    }

    @SuppressWarnings("unchecked")
    private static String nested(Map<String, Object> value, String... path) {
        Object current = value;
        for (String key : path) {
            if (!(current instanceof Map<?, ?> record)) return "";
            current = ((Map<String, Object>) record).get(key);
        }
        return text(current);
    }

    private static List<String> stringList(Object value) {
        List<String> result = new ArrayList<>();
        if (value instanceof List<?> items) for (Object item : items) result.add(text(item));
        return result;
    }

    private static String firstNonEmpty(String... values) {
        for (String value : values) if (!value.isBlank()) return value;
        return "";
    }

    private static String capitalize(String value) {
        return value.isBlank() ? value : value.substring(0, 1).toUpperCase() + value.substring(1);
    }

    private static String url(String value) {
        return URLEncoder.encode(value, StandardCharsets.UTF_8);
    }

    record JiraConfig(String baseUrl, String email, String token) {
        static JiraConfig read() {
            String baseUrl = System.getenv().getOrDefault("JIRA_BASE_URL", "").replaceAll("/+$", "");
            String email = System.getenv().getOrDefault("JIRA_EMAIL", "");
            String token = System.getenv().getOrDefault("JIRA_API_TOKEN", "");
            if (baseUrl.isBlank() || email.isBlank() || token.isBlank()) return null;
            return new JiraConfig(baseUrl, email, token);
        }
    }
}

File: examples/showcase/gitlab_fronting/custom-code-bundles/gitlab_fronting_python/src/gitlab_governed_fronting_showcase/backend_adapter.py
"""GitLab REST backend seam for the governed fronting showcase."""
from __future__ import annotations

import json
import os
import urllib.error
import urllib.parse
import urllib.request
from typing import Any

BackendInvocationPlan = dict[str, Any]
GeneratedCapability = dict[str, Any]


def _gitlab_token() -> str | None:
    return os.getenv("GITLAB_TOKEN", "").strip() or None


def _api_base() -> str:
    return os.getenv("GITLAB_API_BASE", "https://gitlab.com/api/v4").rstrip("/")


def _csv_env(name: str) -> set[str]:
    return {item.strip() for item in os.getenv(name, "").split(",") if item.strip()}


def _project_allowed(project_id: str) -> bool:
    key = project_id.lower()
    blocked = {item.lower() for item in _csv_env("ANIP_GITLAB_BLOCKED_PROJECTS")}
    allowed = {item.lower() for item in _csv_env("ANIP_GITLAB_ALLOWED_PROJECTS")}
    if key in blocked:
        return False
    return not allowed or key in allowed


def _bounded_limit(value: Any, *, default: int = 20, maximum: int = 50) -> int:
    try:
        limit = int(value)
    except (TypeError, ValueError):
        limit = default
    return max(1, min(limit, maximum))


def _list_of_strings(value: Any) -> list[str]:
    if isinstance(value, list):
        raw = value
    elif isinstance(value, str) and value.strip():
        raw = [item.strip() for item in value.split(",")]
    else:
        raw = []
    result: list[str] = []
    for item in raw:
        text = str(item).strip()
        if text and text not in result:
            result.append(text)
    return result


def _project_path(project_id: str) -> str:
    return urllib.parse.quote(project_id, safe="")


def _project_id_from_params(params: dict[str, Any]) -> str:
    explicit = str(params.get("project_id") or "").strip()
    if explicit:
        return explicit
    namespace = str(params.get("namespace") or "").strip().strip("/")
    project = str(params.get("project") or "").strip().strip("/")
    if namespace and project:
        return f"{namespace}/{project}"
    return ""


def _gitlab_request_json(method: str, path: str, token: str, body: dict[str, Any] | None = None) -> Any:
    request = urllib.request.Request(
        f"{_api_base()}{path}",
        data=json.dumps(body).encode("utf-8") if body is not None else None,
        headers={
            "Accept": "application/json",
            "Content-Type": "application/json",
            "PRIVATE-TOKEN": token,
            "User-Agent": "anip-gitlab-fronting-showcase",
        },
        method=method,
    )
    try:
        with urllib.request.urlopen(request, timeout=20) as response:
            raw = response.read().decode("utf-8")
            return json.loads(raw) if raw else {}
    except urllib.error.HTTPError as exc:
        return {
            "error": "gitlab_http_error",
            "status": exc.code,
            "detail": exc.read().decode("utf-8", errors="replace"),
        }


def _approval_grant_from_context(ctx: Any) -> dict[str, Any] | None:
    grant = getattr(ctx, "approval_grant", None)
    if isinstance(grant, dict) and grant.get("grant_id") and grant.get("approval_request_id"):
        return grant
    return None


def _mutation_enabled(ctx: Any) -> bool:
    return os.getenv("ANIP_GITLAB_ALLOW_MUTATION", "").lower() == "true" and _approval_grant_from_context(ctx) is not None


def _metadata(capability: GeneratedCapability, plan: BackendInvocationPlan) -> dict[str, Any]:
    return {
        "capability_id": capability["capability_id"],
        "selected_backend": plan["selected_binding"],
        "semantic_input": plan["semantic_input"],
    }


def _restricted_project_response(capability: GeneratedCapability, plan: BackendInvocationPlan, project_id: str) -> dict[str, Any]:
    return {
        "execution_status": "restricted",
        **_metadata(capability, plan),
        "project_id": project_id,
        "reason": "GitLab project is outside the configured ANIP project policy.",
    }


class DefaultBackendAdapter:
    async def execute(
        self,
        capability: GeneratedCapability,
        plan: BackendInvocationPlan,
        _params: dict[str, Any],
        ctx: Any = None,
    ) -> dict[str, Any]:
        if plan["unresolved_required_backend_inputs"]:
            return {
                "execution_status": "backend_input_incomplete",
                **_metadata(capability, plan),
                "backend_input_contract": plan["backend_input_contract"],
                "unresolved_required_backend_inputs": plan["unresolved_required_backend_inputs"],
            }
        token = _gitlab_token()
        if token is None:
            return {"execution_status": "backend_error", **_metadata(capability, plan), "gitlab_error": {"error": "missing_gitlab_token"}}
        capability_id = capability["capability_id"]
        if capability_id == "gitlab.project.search_context":
            return self._search_project_context(capability, plan, _params, token)
        if capability_id == "gitlab.issue.prepare":
            return self._prepare_or_create_issue(capability, plan, _params, token, ctx)
        if capability_id == "gitlab.mr.comment.prepare":
            return self._prepare_mr_comment(capability, plan, _params, token, ctx)
        if capability_id == "gitlab.pipeline.trigger.request":
            return self._prepare_pipeline_trigger(capability, plan, _params, token, ctx)
        if capability_id == "gitlab.release_notes.prepare":
            return self._prepare_release_notes(capability, plan, _params, token)
        return {
            "execution_status": "backend_execution_stub",
            **_metadata(capability, plan),
            "backend_input_contract": plan["backend_input_contract"],
        }

    def _project_metadata_or_error(
        self,
        capability: GeneratedCapability,
        plan: BackendInvocationPlan,
        params: dict[str, Any],
        token: str,
    ) -> tuple[str, dict[str, Any], dict[str, Any] | None]:
        project_id = _project_id_from_params(params)
        if not _project_allowed(project_id):
            return project_id, {}, _restricted_project_response(capability, plan, project_id)
        project = _gitlab_request_json("GET", f"/projects/{_project_path(project_id)}", token)
        if isinstance(project, dict) and project.get("error"):
            return project_id, {}, {"execution_status": "backend_error", **_metadata(capability, plan), "gitlab_error": project}
        return project_id, project, None

    def _search_project_context(
        self,
        capability: GeneratedCapability,
        plan: BackendInvocationPlan,
        params: dict[str, Any],
        token: str,
    ) -> dict[str, Any]:
        project_id = _project_id_from_params(params)
        if not _project_allowed(project_id):
            return _restricted_project_response(capability, plan, project_id)
        query = str(params.get("query") or "").strip()
        limit = _bounded_limit(params.get("limit"))
        encoded = urllib.parse.urlencode({"search": query, "per_page": str(limit)})
        issues = _gitlab_request_json("GET", f"/projects/{_project_path(project_id)}/issues?{encoded}", token)
        merge_requests = _gitlab_request_json("GET", f"/projects/{_project_path(project_id)}/merge_requests?{encoded}", token)
        if isinstance(issues, dict) and issues.get("error"):
            return {"execution_status": "backend_error", **_metadata(capability, plan), "gitlab_error": issues}
        if isinstance(merge_requests, dict) and merge_requests.get("error"):
            return {"execution_status": "backend_error", **_metadata(capability, plan), "gitlab_error": merge_requests}
        items = [
            {"kind": "issue", "iid": item.get("iid"), "title": item.get("title"), "state": item.get("state"), "web_url": item.get("web_url")}
            for item in list(issues or [])[:limit]
        ]
        remaining = max(0, limit - len(items))
        items.extend(
            {"kind": "merge_request", "iid": item.get("iid"), "title": item.get("title"), "state": item.get("state"), "web_url": item.get("web_url")}
            for item in list(merge_requests or [])[:remaining]
        )
        return {
            "execution_status": "completed",
            **_metadata(capability, plan),
            "gitlab_query": query,
            "result": {"items": items, "count": len(items), "project_id": project_id},
        }

    def _prepare_or_create_issue(
        self,
        capability: GeneratedCapability,
        plan: BackendInvocationPlan,
        params: dict[str, Any],
        token: str,
        ctx: Any,
    ) -> dict[str, Any]:
        project_id, project, error = self._project_metadata_or_error(capability, plan, params, token)
        if error is not None:
            return error
        body: dict[str, Any] = {
            "title": str(params.get("title") or "").strip(),
            "description": str(params.get("body") or params.get("description") or "").strip(),
        }
        labels = _list_of_strings(params.get("labels"))
        if labels:
            body["labels"] = ",".join(labels)
        preview = self._write_preview(capability, plan, "issues.create", f"/projects/{project_id}/issues", body, project)
        if not _mutation_enabled(ctx):
            return preview
        created = _gitlab_request_json("POST", f"/projects/{_project_path(project_id)}/issues", token, body)
        if isinstance(created, dict) and created.get("error"):
            return {**preview, "execution_status": "backend_error", "gitlab_error": created}
        return {
            **preview,
            "execution_status": "completed",
            "approval_required": False,
            "mutation_performed": True,
            "created_issue": {"iid": created.get("iid"), "web_url": created.get("web_url"), "state": created.get("state")},
        }

    def _prepare_mr_comment(self, capability: GeneratedCapability, plan: BackendInvocationPlan, params: dict[str, Any], token: str, ctx: Any) -> dict[str, Any]:
        project_id, project, error = self._project_metadata_or_error(capability, plan, params, token)
        if error is not None:
            return error
        iid = str(params.get("merge_request_iid") or "").strip()
        mr = _gitlab_request_json("GET", f"/projects/{_project_path(project_id)}/merge_requests/{_project_path(iid)}", token)
        if isinstance(mr, dict) and mr.get("error"):
            return {"execution_status": "backend_error", **_metadata(capability, plan), "gitlab_error": mr}
        body = {"body": f"[{params.get('comment_purpose')}] {params.get('context')}".strip()}
        preview = self._write_preview(capability, plan, "merge_requests.createNote", f"/projects/{project_id}/merge_requests/{iid}/notes", body, project)
        preview["merge_request"] = {"iid": mr.get("iid"), "title": mr.get("title"), "state": mr.get("state")}
        return preview

    def _prepare_pipeline_trigger(self, capability: GeneratedCapability, plan: BackendInvocationPlan, params: dict[str, Any], token: str, ctx: Any) -> dict[str, Any]:
        project_id, project, error = self._project_metadata_or_error(capability, plan, params, token)
        if error is not None:
            return error
        body = {
            "ref": str(params.get("ref") or "").strip(),
            "variables": params.get("variables") or {},
            "purpose": str(params.get("pipeline_purpose") or "").strip(),
        }
        return self._write_preview(capability, plan, "pipeline.trigger", f"/projects/{project_id}/pipeline", body, project)

    def _prepare_release_notes(self, capability: GeneratedCapability, plan: BackendInvocationPlan, params: dict[str, Any], token: str) -> dict[str, Any]:
        project_id, project, error = self._project_metadata_or_error(capability, plan, params, token)
        if error is not None:
            return error
        release_range = str(params.get("range") or "").strip()
        return {
            "execution_status": "completed",
            **_metadata(capability, plan),
            "mutation_performed": False,
            "result": {
                "title": f"Release notes for {project.get('path_with_namespace') or project_id} {release_range}",
                "audience": str(params.get("audience") or "internal").strip(),
                "project": self._project_summary(project),
                "range": release_range,
                "sections": [
                    {"title": "Highlights", "items": ["Review bounded GitLab context before publishing release notes."]},
                    {"title": "Governance", "items": ["This capability drafts content only and does not create a GitLab release."]},
                ],
            },
        }

    def _write_preview(self, capability: GeneratedCapability, plan: BackendInvocationPlan, action: str, path: str, body: dict[str, Any], project: dict[str, Any]) -> dict[str, Any]:
        return {
            "execution_status": "prepared",
            **_metadata(capability, plan),
            "approval_required": True,
            "mutation_performed": False,
            "gitlab_action": action,
            "gitlab_metadata": self._project_summary(project),
            "gitlab_request": {"method": "POST", "path": path, "body": body},
            "note": "Prepared a GitLab request payload. No GitLab mutation was performed.",
        }

    def _project_summary(self, project: dict[str, Any]) -> dict[str, Any]:
        return {
            "id": project.get("id"),
            "path_with_namespace": project.get("path_with_namespace"),
            "default_branch": project.get("default_branch"),
            "visibility": project.get("visibility"),
            "web_url": project.get("web_url"),
        }


backend_adapter = DefaultBackendAdapter()
